File: 알고리즘 정리/BFS,DFS/문제풀이/네트워크.cpp
#include <iostream>

//////////////////////////////////////////////////////////
#include <string>
#include <vector>
#include <queue>
using namespace std;
vector<int> connected[201];

int solution(int n, vector<vector<int>> computers) 
{
	int answer = 0;
	for (int i = 0; i < n; i++)
		for (int j = 0; j < n; j++)
			if (i != j&&computers[i][j]==1)
				connected[i].push_back(j);

	bool visited[201] = { 0 };
	for (int i = 0; i < n; i++)
	{
		if (!visited[i])
		{
			++answer;

			//bfs
			queue<int> q;
			q.push(i);
			visited[i] = 1;

			while (!q.empty())
			{
				int x = q.front();
				q.pop();
				for (int j = 0; j < connected[x].size(); j++)
				{
					int y = connected[x][j];
					if (!visited[y])
					{
						visited[y] = 1;
						q.push(y);
					}
				}
			}
			//end bfs
		}
	}

	return answer;
}

//////////////////////////////////////////////////////////


int main() 
{
	cin.sync_with_stdio(0);
	cin.tie(0); cout.tie(0);
	int n = 3;
	vector<vector<int>> computers = { {1,1,0},{1,1,0},{0,0,1} };
	cout << solution(n, computers) << '\n';
	return 0;
}

File: 알고리즘 정리/BFS,DFS/문제풀이/네트워크.py
################################################
from collections import deque
def solution(n, computers):
    answer = 0
    visited = [0 for _ in range(201)]
    q = deque()
    for i in range(n):
        if visited[i] == 0:
            answer += 1
            visited[i] = 1
            q.appendleft(i)
            while q:
                x = q.pop()
                for i in range(n):
                    if i != x and computers[x][i] == 1 and visited[i] == 0:
                        q.appendleft(i)
                        visited[i] = 1

    return answer

################################################

n=3
computers=[[1, 1, 0], [1, 1, 0], [0, 0, 1]]
print(solution(n,computers))